### PR TITLE
New attribute based input system

### PR DESCRIPTION
**Pull request by FuJa0815**
I had an implementation of an attribute based input system in this PR: #1697, but it had many problems and bugs.
This implementation is cleaner and more structured.

- [x] Fix merge conflicts
- [x] Make InputAttribute remove dead weak references once per second
- [x] need to do some overhauling on suppressing other inputs after some key reports being used (especially in axis groups)? (I don't fully remember what this was about...). I remember now: this is about disposed exceptions coming from the editor when returning to the stage.
- [x] axis keys need to go back to that latest pressed input overrides others mode
- [x] some stuff need _Input while most other stuff need unhandled input, Priority property in the input attributes that switches between them?
      These should use priority input: TransitionManager.cs, QuickLoadHandler.cs, ScreenShotTaker.cs, FPSCounter.cs
- [x] I think all `InputManager.RegisterReceiver(this);` calls should be in tree enter and unregister in tree exit? SO that objects that are in the background and not attached, don't receive input accidentally. Needs updating the example docs. For example check does microbe camera get input even when not active as in the case of being in the editor
- [x] check that the rename: InvokeWithNoInput -> InvokeAlsoWithNoInput is done
- [x] OPTIONAL: recreate InputToggle for not needing to handle local state in some input states, would work like other key input but it was internally tracked if the key was toggled on or off and that was passed to the method
- [x] OPTIONAL: more logic should be rather in InputManager than in InputAttribute
- [x] OPTIONAL: zoom in speed disparency between mouse wheel and keypad + and - with new input system
- [ ] OPTIONAL: make InvokeAlsoWithNoInput not run (put a breakpoint in RunOnAxisGroupAttribute.OnProcess and see it trigger before even starting the game) in unrelated gamestate somehow, or solve that threading issue so that we have enough performance to just repeatedly call useless code each frame. This is at least partly solved by that move the register / unregister calls to tree enter and exit.

### CODE REVIEW

**Comment by FuJa0815** [comment]
I'll revert the new input system in #1697 once this is merged.

File: src/engine/input/RunOnAxisAttribute.cs
Comment (Kasterisk) [inline src/engine/input/RunOnAxisAttribute.cs:74]:
```suggestion
        var currentResult = CurrentResult;
```

File: src/engine/input/RunOnAxisAttribute.cs
Comment (Kasterisk) [inline src/engine/input/RunOnAxisAttribute.cs:18]: Should be multiple lines here too

File: src/engine/input/InputManager.cs
Comment (Kasterisk) [inline src/engine/input/InputManager.cs:16]: Summary should be multiple lines

File: src/engine/input/RunOnAxisAttribute.cs
Comment (Kasterisk) [inline src/engine/input/RunOnAxisAttribute.cs:23]:
```suggestion
    ///   Instantiates a new RunOnAxisAttribute.
```

File: src/engine/input/RunOnAxisGroupAttribute.cs
Comment (Kasterisk) [inline src/engine/input/RunOnAxisGroupAttribute.cs:27]:
```suggestion
    public bool InvokeWithNoInput { get; set; }
    
    public override bool OnInput(InputEvent @event)
```

File: project.godot
Comment (hhyyrylainen) [inline project.godot:33]: Input manager should be after startupactions.

File: src/tutorial/microbe_stage/MicrobeMovement.cs
Comment (hhyyrylainen) [inline src/tutorial/microbe_stage/MicrobeMovement.cs:93]: Is this attribute needed? I think it would be simpler to use the system if specifying RunOnAxis attributes would automatically make the input manager handle this as if RunOnAxisGroup was specified.

Comment (hhyyrylainen) [reply src/tutorial/microbe_stage/MicrobeMovement.cs:93]: I see `[RunOnAxisGroup(InvokeWithNoInput = true)]` is used elsewhere, but when not needing any special settings, I think that it would be easier for future programmers if the input system assumed that RunOnAxis means default RunOnAxisGroup if that attribute is not specifically added.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: src/general/PauseMenu.cs
Comment (Kasterisk) [inline src/general/PauseMenu.cs:78]:
```suggestion
    public void EscapeKeyPressed()
```

File: src/engine/input/InputAttribute.cs
Comment (Kasterisk) [inline src/engine/input/InputAttribute.cs:14]: Use xml summary here

File: src/engine/input/InputAttribute.cs
Comment (Kasterisk) [inline src/engine/input/InputAttribute.cs:17]: And xml summary here too

File: src/engine/input/RunOnKeyAttribute.cs
Comment (Kasterisk) [inline src/engine/input/RunOnKeyAttribute.cs:15]: I think this could use xml style

**Review by Kasterisk — DISMISSED** [review]

File: src/tutorial/microbe_stage/MicrobeMovement.cs
Comment (FuJa0815) [reply src/tutorial/microbe_stage/MicrobeMovement.cs:93]: Without the `RunOnAxisGroup` this would mean, that this Method has two `RunOnAxisAttribute`s attached to it. When either one of those axes is triggered, the method will be called with `(float delta, float value)`. With `RunOnAxisGroup` the system knows that theese represent different axes and therefore call `(float delta, float value1, float value2)`.
In a nutshell: you can attach multiple `InputAttributes` to a method. Therefore it is valid to attach multiple `RunOnKeyDownAttribute`s or multiple `RunOnAxisAttribute`s. The problem is, that you then can't differenciate which key or axes triggered this method.

It is very unlikely that this will ever be utilized, but the system supports it. I hope you get what I mean

Comment (FuJa0815) [reply src/tutorial/microbe_stage/MicrobeMovement.cs:93]: One example for multiple InputAttributes would be, if you had
- slow zoom in
- slow zoom out
- fast zoom in
- fast zoom out

These two implementations would be valid:
```
[RunOnAxis(new[] { "g_zoom_out_slow", "g_zoom_in_slow", "g_zoom_out_fast", "g_zoom_in_fast" }, new[] { -1, 1, -3, 3 })]
```

```
[RunOnAxis(new[] { "g_zoom_out_slow", "g_zoom_in_slow" }, new[] { -1, 1 })]
[RunOnAxis(new[] { "g_zoom_out_fast", "g_zoom_in_fast" }, new[] { -3, 3 })]
```

Comment (hhyyrylainen) [reply src/tutorial/microbe_stage/MicrobeMovement.cs:93]: I now seem to remember something about that. Could you write more xml comments, and maybe write a short overview like "doc/input_system.md" file. This is a pretty important system for new programmers *and* me to understand.

**Comment by FuJa0815** [comment]
I've added a [input_system.md](https://github.com/Revolutionary-Games/Thrive/blob/1e4ce413bcb680b3cc2b196ea40d7b0f6d5380df/doc/input_system.md) file. English is not my native language, therefore somebody should spellcheck and give feedback if the content is understandable

**Comment by FuJa0815** [comment]
Thanks @Nunez2196!

Is the content understandable?

**Comment by Nunez2196** [comment]
@FuJa0815 Yes, it reads well, there were just some misspelled words was all I saw

File: project.godot
Comment (hhyyrylainen) [inline project.godot:214]: Looks like you deleted a bunch of locales, probably by accident.

File: src/tutorial/microbe_stage/MicrobeMovement.cs
Comment (hhyyrylainen) [inline src/tutorial/microbe_stage/MicrobeMovement.cs:94]: I didn't notice this before, but because this is in the *tutorial* and because of the way I wrote the tutorial system. This should be left as is, and not be converted to the new input system.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: src/microbe_stage/PlayerMicrobeInput.cs
Comment (hhyyrylainen) [inline src/microbe_stage/PlayerMicrobeInput.cs:35]: Why is this sent to the input manager? Why doesn't input manager implement _Nofitication?

File: src/microbe_stage/MicrobeCamera.cs
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeCamera.cs:129]: Should delta be multiplied into "value" here because otherwise the camera zoom speed is framerate dependent?

File: src/gui_common/QuickLoadHandler.cs
Comment (hhyyrylainen) [inline src/gui_common/QuickLoadHandler.cs:26]: _Notification also used here to notify InputManager.

File: src/general/PauseMenu.cs
Comment (hhyyrylainen) [inline src/general/PauseMenu.cs:78]: I think the new input system should handle priority between this and transition manager. Godot inbuilt system with the other place calling SetInputAsHandled() properly only allows the top one of these two objects to succeed. Maybe RunOnKeyDown needs a new Exclusive attribute which stops event propagation when it is fired and the invoked method returns true (so this type of invoke would need to return a boolean, I suppose if the return type is void it can be assumed that the input is always consumed and unavailable to other input receivers).

File: src/engine/input/RunOnKeyUpAttribute.cs
Comment (hhyyrylainen) [inline src/engine/input/RunOnKeyUpAttribute.cs:9]: I don't think using `godotInputName` is necessary, it should just be `inputName` if we ever have some input name that doesn't match a Godot action, that should instead have a more complicated name.

File: src/microbe_stage/MicrobeCamera.cs
Comment (FuJa0815) [reply src/microbe_stage/MicrobeCamera.cs:129]: I've kept the old system, which also was framerate dependent

Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeCamera.cs:129]: It's good time to fix this now.

Comment (hhyyrylainen) [reply src/microbe_stage/MicrobeCamera.cs:129]: I think, assuming about 60 FPS, you an equivalent formula would be `value * delta * 55` 55 is from calculating how much you need to multiple 18 milliseconds by to get back to 1 second.

File: src/general/PauseMenu.cs
Comment (FuJa0815) [reply src/general/PauseMenu.cs:78]: Do you mean something like https://github.com/Revolutionary-Games/Thrive/pull/1867/commits/91417cdf3c17442008d96fcb2f922177e715ff91 and https://github.com/Revolutionary-Games/Thrive/pull/1867/commits/d5ddfdf9202d88cf168d6b45e182ce37384bd952?

**Comment by FuJa0815** [comment]
One more thing: I've noticed a crash when skipping the Microbe Stage intro video. The crash is very weird, because the game just closes and my IDE doesn't even detect an exeption. The game crashed on this line: https://github.com/Revolutionary-Games/Thrive/blob/eef35b846f9b1d32b99d5cc5a76068648e420161/src/engine/SceneManager.cs#L46
surrounding this line with a try catch also just closes the game... weird...

**Comment by hhyyrylainen** [comment]
Godot doesn't throw exceptions for most operations, it just only logs an error *or* on bad errors it just closes the game.

File: src/general/PauseMenu.cs
Comment (hhyyrylainen) [reply src/general/PauseMenu.cs:78]: From looking at those it seems you now have the part that the method can tell that it did or did not consume the key press. Did you already have the code for skipping any further callbacks once one consumes the input?

**Comment by FuJa0815** [comment]
Regarding the `AsParallel()` thing:
I've written this small code:
```csharp
            var s = new Stopwatch();
            float sumSingleThread = 0.0f;
            float sumMultiThread = 0.0f;
            for (int i = 0; i < 100; i++)
            {
                s.Reset();
                s.Start();
                Enumerable.Range(0, 50000000).ToList().ForEach(p => _ = p);
                sumSingleThread += s.ElapsedMilliseconds;
                s.Reset();
                s.Start();
                Enumerable.Range(0, 50000000).ToList().AsParallel().ForAll(p => _ = p);
                sumMultiThread += s.ElapsedMilliseconds;
            }
            Console.WriteLine(sumSingleThread/100.0f);
            Console.WriteLine(sumMultiThread / 100.0f);
```
which printed out
```
267.75
234.66
```
on my mashine (Intel i7-6700K @ 4.00GHz 8 Cores)

**Comment by hhyyrylainen** [comment]
That shows what the C# documentation says about ForAll is true, that it is executed in parallel.
It's just that, as I wrote in this issue: https://github.com/Revolutionary-Games/Thrive/issues/1874, we are making a game and need to be very thoughtful about the number of threads running at once, and also we need to be very careful about what Godot operations are called from background threads. For those reasons adding threading to input handling needs careful design and a lot of testing.

**Comment by hhyyrylainen** [comment]
It's once again taking a bit of time before I can find time to work on this more, here's my notes I have so far for what needs fixing:

rename: InvokeWithNoInput -> InvokeAlsoWithNoInput
need to test skipping microbe intro video (why does it crash?)
why is CallMethod(values.Cast<object>().ToArray()); needed
some stuff need _Input while most other stuff need unhandled input, Priority property to input attributes? And then add new parallel execution path for priority input handling
make the average value be rounded to ensure no 0.0000001f or such values can be there

**Comment by Kasterisk** [comment]
I think I found a bug, the key prompt for the cell movement tutorial doesn't match up with the key press. For instance, if the player is pressing S key, the W on the key prompt is the one hidden instead.

**Comment by hhyyrylainen** [comment]
I've basically reverted the tutorial related code back to the previous state, I haven't committed yet as I'm going to make one commit fixing all the small problems at once.

**Comment by FuJa0815** [comment]
The `.Cast<object>().ToArray()` is required to fulfill the `params object[] parameters`.
Without the Cast the CallMethod's parameters would contain one float[], but I want parameters to contain the individual float values. Thats why I have to call .Cast to match the params parameter.

**Comment by hhyyrylainen** [comment]
One thing that is gone with this new system is the nice handling of the input axis keys, now this doesn't work:
- Press left
- Cell moves left
- Press right (without releasing anything)
- Cell doesn't move at all! Cell should move right instead, with my previous code it worked
- Release right
- Cell goes left again (at least this part works still).

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

**Comment by hhyyrylainen** [comment]
A few more problems that need fixing that I don't have time for:
- Add new comment for this https://github.com/Revolutionary-Games/Thrive/pull/1867#discussion_r533563296
- Background objects receiving input is a major problem, I fixed the crash but it spams logs a lot regarding the camera. The register and unregister for input need to be moved to enter tree and exit tree for non-autoload classes.
- TransitionManager.cs, QuickLoadHandler.cs, ScreenShotTaker.cs, FPSCounter.cs need that priority _Input receiving that can't be blocked by any other input receivers

And one thing I didn't test yet is saving, especially saving in the editor.

**Comment by FuJa0815** [comment]
> 
> 
> One thing that is gone with this new system is the nice handling of the input axis keys, now this doesn't work:
> 
>     * Press left
> 
>     * Cell moves left
> 
>     * Press right (without releasing anything)
> 
>     * Cell doesn't move at all! Cell should move right instead, with my previous code it worked
> 
>     * Release right
> 
>     * Cell goes left again (at least this part works still).

I intentionally implemented the input system to behave exactly like this. I thought the old behaviour was a bug! In my opinion pressing left and right should result in no movement.

**Comment by hhyyrylainen** [comment]
I like the approach where the last pressed key has precedence. Usually it has no effect, but for example strafing in an FPS setting would be faster if you don't need to manage to let go of the other direction first. If no one else agrees with me, I'm okay leaving it as is, but it'd be really cool to have built to my liking input in thrive (and preserve what the previous code already did).

**Comment by hhyyrylainen** [comment]
I compiled a list of the things still needing doing. I put that in the PP description (top of this page).

**Comment by FuJa0815** [comment]
Is the failure of format_script the same problem as here? https://github.com/Revolutionary-Games/Thrive/pull/1772#issuecomment-740959469

**Comment by hhyyrylainen** [comment]
Maybe if you haven't updated? Also you need to definitely update Babel-Thrive as I did some big changes to that.

**Comment by FuJa0815** [comment]
I have updated everything (git pull, apt update/upgrade), ran the localization script again and nothing changed, except the POT-Creation-Date.

**Comment by hhyyrylainen** [comment]
You need to give pip a special flag to install updates. Or you can use the new requirements file I added with pip to make sure it installs the right versions, see the updated setup instructions: https://github.com/Revolutionary-Games/Thrive/blob/master/doc/setup_instructions.md#babel-and-babel_thrive

File: assets/models/EditorArrow.tscn
Comment (hhyyrylainen) [inline assets/models/EditorArrow.tscn:1]: This PR seems to have been mixed up with the editor arrow one.
You should do the following: create a new local branch from this (this in effect saves this branch with a new name), delete this branch locally, create this branch again from master, and then cherry-pick all the right commits from the new branch where you saved the state of this branch

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: locale/messages.pot
Comment (hhyyrylainen) [inline locale/messages.pot:8]: Could you revert all the translations (checkout the files from latest master commit), as this PR doesn't add any new translation strings, so this doesn't need to contain any translation changes?

File: src/engine/ScreenShotTaker.cs
Comment (hhyyrylainen) [inline src/engine/ScreenShotTaker.cs:8]: I'll do a full review later, but this name started bothering me a few days ago when I saw it.
I think "InputNode" give the wrong connotation, that this is just something for input handling. I'd prefer a name like NodeWithInput, that way it is clear that this is an expanded Node class "with input".

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
https://github.com/Revolutionary-Games/Thrive/pull/1867/files#r542506818
https://github.com/Revolutionary-Games/Thrive/pull/1867/files#r542506063

File: doc/input_system.md
Comment (hhyyrylainen) [inline doc/input_system.md:31]: Shouldn't this be "Fires when the state of the toggle changes"
Because just above "pressed" seems to mean that the key is pressed down, so this is misleading documentation.

File: doc/input_system.md
Comment (hhyyrylainen) [inline doc/input_system.md:46]: Still uses the old name here

File: doc/input_system.md
Comment (hhyyrylainen) [inline doc/input_system.md:44]: "you must do one of the following:"

File: doc/input_system.md
Comment (hhyyrylainen) [inline doc/input_system.md:57]: This is outdated naming `InvokeAlsoWithNoInput`

File: doc/input_system.md
Comment (hhyyrylainen) [inline doc/input_system.md:51]: Lines under sections should be the same length as the text of the title.

File: doc/input_system.md
Comment (hhyyrylainen) [inline doc/input_system.md:95]: Outdated naming is also used here in the examples

File: doc/input_system.md
Comment (hhyyrylainen) [inline doc/input_system.md:188]: outdated naming

File: src/engine/input/ControlWithInput.cs
Comment (hhyyrylainen) [inline src/engine/input/ControlWithInput.cs:4]: This documentation should be more comprehensive. This should say what being an "input class" means.

File: src/engine/input/NodeWithInput.cs
Comment (hhyyrylainen) [inline src/engine/input/NodeWithInput.cs:4]: Or maybe it would be better to have more comprehensive documentation here, and then link to this from ControlWithInput.

File: src/engine/input/InputManager.cs
Comment (hhyyrylainen) [inline src/engine/input/InputManager.cs:47]: This shouldn't touch the internal state of the attribute like this.

File: src/engine/input/InputManager.cs
Comment (hhyyrylainen) [inline src/engine/input/InputManager.cs:22]: I think a good way to change this would be to change this into a dictionary like: `Dictionary<InputAttribute, List<WeakReference>>` in order to store the instances related to each input attribute. That way this class owns the instance references so this doesn't have to violate encapsulation by directly accessing InputAttribute.Instances.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]

File: src/engine/input/InputManager.cs
Comment (hhyyrylainen) [inline src/engine/input/InputManager.cs:186]: I'd prefer this to be a written out loop (instead of LINQ) just because the ordering matters because of the input consuming.

File: src/engine/input/InputManager.cs
Comment (hhyyrylainen) [inline src/engine/input/InputManager.cs:209]: This should be named `ClearExpiredReferences`

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
Just two more really tiny things.

**Review by hhyyrylainen — APPROVED** [review]
Finally done with the final set of changes I wanted to do.
Fixed two bugs:
- zoom speed was inconsistent
- There was a crash if you clicked the mouse after returning from the editor

File: src/engine/input/InputManager.cs
Comment (FuJa0815) [inline src/engine/input/InputManager.cs:227]: Wouldn't it make more sense to change this to isDown?

Comment (FuJa0815) [reply src/engine/input/InputManager.cs:227]: I mean inverting the variable.
````suggestion
        bool isDown = @event.IsPressed();
```

**Pull request merged**
